test/src/test/java/hudson/model/QueueTest.java

```diff
@@ -707,5 +709,38 @@ public void run() {
             fail("Expected an CancellationException to be thrown");
         } catch (CancellationException e) {}
     }
-
+    
+    @Issue("JENKINS-27871")
+    @Test public void testBlockBuildWhenUpstreamBuildingLock() throws Exception {
+        final String prefix = "JENKINS-27871";
+        r.getInstance().setNumExecutors(4);
+        r.getInstance().save();
+        
+        final FreeStyleProject projectA = r.createFreeStyleProject(prefix+"A");
+        projectA.getBuildersList().add(new SleepBuilder(5000));
+        
+        final FreeStyleProject projectB = r.createFreeStyleProject(prefix+"B");
+        projectB.getBuildersList().add(new SleepBuilder(10000));     
+        projectB.setBlockBuildWhenUpstreamBuilding(true);
+        
+        final FreeStyleProject projectC = r.createFreeStyleProject(prefix+"C");
+        projectC.getBuildersList().add(new SleepBuilder(10000));
```

Comment:
This forces the test to take at least 10s, but it is still going to fail randomly on a heavily loaded test machine. Would be better to use a semaphore-based builder. (Not sure offhand if we have a standard one lying around.)